### ISSUE TITLE
Rule Change: make `complexity` less visually disruptive

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#19984]
### What rule do you want to change?

complexity

### What change do you want to make?

The rule is fine, I'm not asking to change the behavior, but just its _look_.

It's completely disruptive and this leads to developers disabling the rule inline simply to be able to work on the affected function. The rule, even set as a warning, completely shadows any and every other warning in the function.

This can be done by _selecting_ `FunctionDeclaration > Identifier` instead of `FunctionDeclaration`. The only unknown here is what to do with anonymous functions. Maybe highlight `function` or `=>`.

### What does the rule currently do for this code?

<img width="696" height="376" alt="Image" src="https://github.com/user-attachments/assets/edef7a52-8aab-4849-8aea-4defd2195c45" />

### What will the rule do after it's changed?

<img width="696" height="376" alt="Image" src="https://github.com/user-attachments/assets/827f238c-f264-4e40-8b24-2a88c6e38fe7" />

### Participation

I am potentially willing to submit a pull request to implement this change.

**Comment by lumirlumir** [comment]
I think this is a good approach, as highlighting all of the lines could lead to unclear guidance.

I’m personally 👍 on this suggestion, but before moving forward, I’d like to gather a few more thoughts from the team. @eslint/eslint-team

**Comment by JoshuaKGoldberg** [comment]
Agreed on the problem statement. Rules that report on an entire class, function, or other large area/block are rather disruptive. They drown out other complaints and are visually intimidating at first. 

However, I'm not sure what the best path forward is. Underlining just the function name is a pretty small surface area, but isn't actually code from the reported issue! Maybe having the report range include the `function` keyword & any other modifier(s) such as `export`, as well as the parameter, would be a good compromise?

```ts
export function getToggleOptions({
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  one,
~~~~~~
  two,
~~~~~~
}) {
~~~~
  // ...
  // ...
  // ...
}
```

**Comment by fregante** [comment]
> pretty small surface area

Is that really an issue? There are rules that highlight two characters (`?.`) so highlighting exactly the identifier here sounds appropriate, regardless of number of characters.

**Comment by nzakas** [comment]
@fregante can you put together a proposal for what should be highlighted for all the different kinds of functions:

1. Function declarations
2. Function expressions (can be anonymous)
3. Arrow function expressions (always anonymous)
4. Object methods
5. Class methods

I'd suggest looking through other function-related rules to see how they handle these situations.

**Comment by fregante** [comment]
> I'd suggest looking through other function-related rules to see how they handle these situations.

From a quick look, the behavior described here is the standard:

- https://eslint.org/docs/latest/rules/func-name-matching
- https://eslint.org/docs/latest/rules/func-style
- https://eslint.org/docs/latest/rules/max-lines-per-function
- https://eslint.org/docs/latest/rules/max-statements

With this being a more moderate highlight:

- https://eslint.org/docs/latest/rules/func-names

And finally this matching what I want `complexity` to look like:

- https://eslint.org/docs/latest/rules/max-params

So maybe I'd suggest extracting a `reportFunction` utility from `max-params` and reuse it in `complexity` (and other rules that are similarly noisy)

**Comment by fregante** [comment]
Yep, this is pretty much what I had in mind:

1. highlight name
2. highlight `function`
3. highlight `=>`





[<img width="476" height="845" alt="Image" src="https://github.com/user-attachments/assets/bd39ffab-a292-485b-aa26-c4c33e780192" />](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbWF4LXBhcmFtczogW1wiZXJyb3JcIiwgM10qL1xuXG5mdW5jdGlvbiBkZWNsYXJhdGlvbihiYXIsIGJheiwgcXV4LCBxeHgpIHtcbiAgZG9Tb21ldGhpbmcoKTtcbn1cblxubGV0IGFycm93ID0gKGJhciwgYmF6LCBxdXgsIHF4eCkgPT4ge1xuICBkb1NvbWV0aGluZygpO1xufTtcblxuY29uc3Qgb2JqID0ge1xuICByZWd1bGFyOiBmdW5jdGlvbiAoYmFyLCBiYXosIHF1eCwgcXh4KSB7XG4gICAgZG9Tb21ldGhpbmcoKTtcbiAgfSxcbiAgYXJyb3c6IChiYXIsIGJheiwgcXV4LCBxeHgpID0+IHtcbiAgICBkb1NvbWV0aGluZygpO1xuICB9LFxuICBzaG9ydEhhbmQoYmFyLCBiYXosIHF1eCwgcXh4KSB7XG4gICAgZG9Tb21ldGhpbmcoKTtcbiAgfSxcbn07XG5cbmNsYXNzIFgge1xuICBtZXRob2QoYmFyLCBiYXosIHF1eCwgcXh4KSB7XG4gICAgZG9Tb21ldGhpbmcoKTtcbiAgfVxuICBib3VuZCA9IChiYXIsIGJheiwgcXV4LCBxeHgpID0+IHtcbiAgICBkb1NvbWV0aGluZygpO1xuICB9O1xufVxuXG5jb25zb2xlLmxvZygoYmFyLCBiYXosIHF1eCwgcXh4KSA9PiB7XG4gIGRvU29tZXRoaW5nKCk7XG59KTtcblxuY29uc29sZS5sb2coZnVuY3Rpb24gKGJhciwgYmF6LCBxdXgsIHF4eCkge1xuICBkb1NvbWV0aGluZygpO1xufSk7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6e319fX19)

**Comment by nzakas** [comment]
Thanks for digging into that. I agree, I think those are good patterns and match what we already have. Let's do it!

**Comment by Un5had0wed** [comment]
@nzakas , @fregante 

Hey guys, is anyone working on this ?? If not, I am interested

**Comment by fregante** [comment]
Please go ahead 🙏

**Comment by ghost** [comment]
@Un5had0wed I am working on this

**Comment by Un5had0wed** [comment]
> [@Un5had0wed](https://github.com/Un5had0wed) I am working on this

Aaah...The issue was unclaimed 🥲. Did i overlook it ?

**Comment by ghost** [comment]
I’m not sure 😅. This looked like a good first issue, so I picked it up.

**Comment by Un5had0wed** [comment]
@yarbfooz Is it okay if i take this issue ??

**Comment by ghost** [comment]
If I’m not able to finish it, I’ll let you know 🙏

**Comment by Un5had0wed** [comment]
Ok. Go ahead then.

**Comment by lumirlumir** [comment]
Just a heads up, according to our [contributing guidelines](https://eslint.org/docs/latest/contribute/work-on-issue#starting-work), we generally go with the person who first claims the issue to work on :)

<img width="881" height="558" alt="Image" src="https://github.com/user-attachments/assets/9d945393-d947-4669-bf1d-8a89ddb1d3a1" />

**Comment by nzakas** [comment]
Just to recap: this issue was claimed originally by @fregante (you can see his name in the "Assignees" list on the issue). @fregante then gave @Un5had0wed [permission](https://github.com/eslint/eslint/issues/19984#issuecomment-3227178576) to take over work on the issue.
